default mapping: Y => y$

teto opened the pull request:
`nnoremap Y y$`

janlazo commented on runtime/plugin/nvim.vim:
```diff
@@ -0,0 +1 @@
+nnoremap Y y$
```
Can the user skip this?

jamessan commented:
There was previous discussion about this [here](https://github.com/neovim/neovim/issues/416).  The issue I foresee with dumping this in a plugin is what @janlazo's hinting at -- there's not an obvious way to skip it.

One would need to setup an `after/` override or `VimEnter` autocmd, which both feel heavy-weight.

teto commented:
all good points. we could map it only if it has not been mapped already (which I find ugly). Alternatively change the C code.

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -189,6 +189,7 @@ Input/Mappings:
 
 Normal commands:
   |gO| shows a filetype-defined "outline" of the current buffer.
+  Y now behaves as y$
```
we need a new section for `Default mappings:` .

Also weird to say "behaves as", that is vague for no reason. We would instead say "is mapped to", but don't need to since it will be clear by its presence in the `Default mappings:` section.

justinmk reviewed (commented):
not sure why we would change the `nv_optrans` implementation which breaks plugins for no reason. 

Instead, we can set a default mapping in main(). Default mappings will be a nice concept to establish, because there are a few other ones we want, such as `CTRL-U`.

> There was previous discussion about this [here](https://github.com/neovim/neovim/issues/416). The issue I foresee with dumping this in a plugin is what @janlazo's hinting at -- there's not an obvious way to skip it.

We can create a `init_default_mappings()` function that happens before startup / user config. Then users can just unmap it if they don't want it.

The main hesitancy was that some plugins check for existing maps to decide things. But `init_default_mappings()`  is the least-bad way to do it. So let's just do it.

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -72,6 +72,12 @@ the differences.
 - |g:vimsyn_embed| defaults to "l" to enable Lua highlighting
 
+==============================================================================
+3. Nvim default Mappings                              *nvim-default-mappings*
```
don't need to mention "Nvim" in a section of a doc where every line in the doc is relevant to Nvim.
```suggestion
3. Default Mappings                              *nvim-default-mappings*
```

justinmk commented on runtime/doc/change.txt:
```diff
@@ -985,10 +985,7 @@ inside of strings can change!  Also see 'softtabstop' option. >
-			yy, |linewise|).  If you like "Y" to work from the
-			cursor to the end of line (which is more logical,
-			but not Vi-compatible) use ":map Y y$".
+["x]Y			Yank [count] lines from the cursor to the end of line.
```
The doc for `:help Y` would not change, rather the pattern would be something like this: 

1. mention the default mapping next to the (legacy) builtin.
2. add a tag like `*Y-default*`

```
*Y-default*
This command is mapped to "y$" by default. |nvim-default-mappings|
```

Commit:
feat(defaults): map Y to y$

rationale:
- consistent with D and Y
- long recommended by Vim's own ":help Y"

close #13268
close #416
ref #6289

justinmk commented on runtime/doc/change.txt:
```diff
@@ -986,10 +986,10 @@ inside of strings can change!  Also see 'softtabstop' option. >
-			cursor to the end of line (which is more logical,
-			but not Vi-compatible) use ":map Y y$".
-
+			yy, |linewise|). This is the builtin mapping, see
+			|Y-default| for neovim's default mapping.
```
- the builtin normal command is not a "mapping", it is a (normal) command.
- we never say "neovim" (lowercase) in the help docs or anywhere else, and also rarely "Neovim". Almost always the help docs say "Nvim". In this case there isn't a reason to mention Nvim at all.
    - I think we need a linter for this at this point...
- There is really no reason to say anything at all here, given that the very next line mentions the default mapped nature.
    - Might seem trivial, but we need a repeatable pattern that doesn't result in clunky text all over the docs.

```suggestion
			yy, |linewise|).
```

justinmk commented on runtime/doc/change.txt:
```diff
@@ -986,10 +986,10 @@ inside of strings can change!  Also see 'softtabstop' option. >
+			yy, |linewise|). This is the builtin mapping, see
+			|Y-default| for neovim's default mapping.
+							*Y-default*
+["x]Y                   This command maps to "y$". see |nvim-default-mappings|
```
- I do not understand why the `["x]Y` in column 1 is repeated.
- Don't need "see".
- Can be more terse without loss of clarity, and now we have a repeatable pattern.

```suggestion
["x]Y                   Maps to "y$" by default. |nvim-default-mappings|
```

justinmk commented on src/nvim/main.c:
```diff
@@ -314,6 +314,9 @@ int main(int argc, char **argv)
   init_highlight(true, false);  // Default highlight groups.
   TIME_MSG("init highlight");
 
+  init_default_mappings();  // Default highlight groups.
```
```suggestion
  init_default_mappings();  // Default mappings.
```

justinmk commented:
pushed some changes:

- move init_default_mappings() to normal.c
- revert all of the random test changes, do `unmap Y` in setup.vim

zeertzjq commented:
Is mapping `Y` to `y$` in visual mode necessary? It makes `v_Y` yank the highlighted text instead of lines, then move cursor to the end of line.

justinmk commented:
> Is mapping `Y` to `y$` in visual mode necessary? It makes `v_Y` yank the highlighted **text** instead of lines, then move cursor to the end of line.

Yeah I should have thought about that more.

spywhere commented:
Hmm, I'm wondering why it's `nnoremap Y y$` over `nnoremap Y yg_`? As the latter would ignore carriage return when yanked.

zeertzjq commented:
> Hmm, I'm wondering why it's `nnoremap Y y$` over `nnoremap Y yg_`? As the latter would ignore carriage return when yanked.

Doesn't `y$` also ignore carriage return? Plus, `yg_` also ignores trailing whitespace, while `D` does delete trailing whitespace.

gpanders commented on src/nvim/normal.c:
```diff
@@ -397,6 +397,11 @@ void init_normal_cmds(void)
+{
+  add_map((char_u *)"Y y$", NORMAL | VISUAL);
+}
+
```
@justinmk I know you moved this from `main.c` to `normal.c`, but for non-normal mode default mappings (e.g. `<C-U>` in insert mode) does it make sense to move this back into `main.c` or should we just put all defaults here in `normal.c`?

justinmk replied:
i guess technically it should go in getchar.c

huyz commented:
I started using neovim a few months ago and I initially welcomed this new mapping default. Loved the consistency.
After 3 months of use, I've gotten very used to hitting `yy` when I want to yank a line.

But now I can see why `Y` is the way it is for vi and vim: ergonomics.
`Y` is used incredibly often and `y$` is comparatively rarely needed.
The problem is that at the speed that I type, I mis-press `yy` quite often (only a single `y` gets registered) on my Keychron keyboard.  I never had that problem with `Y`. 

So yeah, I'm going back to the original meaning of `Y` and I command the original vi author for their counterintuitive wisdom.

Perhaps the way to look at it is that it's `D` that is inconsistent?

Not going to bother creating a separate issue because I may be the only lone voice in this. Still I had to throw that out there for people to consider.

justinmk commented:
> `Y` is used incredibly often and `y$` is comparatively rarely needed.

Nonsense.

huyz commented:
> Nonsense

Wanna bet?

huyz commented:
> > `Y` is used incredibly often and `y$` is comparatively rarely needed.
> 
> Nonsense.

Ah, what I meant is that: yanking a line is needed incredibly often and yanking to end of the line is comparatively rarely needed.

bfredl commented:
No need to over-analyze. you can already remap things to be consistent with your own sense of ergonomics.
